src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -680,23 +686,152 @@ public new CSharpCompilation AddSyntaxTrees(IEnumerable<SyntaxTree> trees)
+            ref ImmutableDictionary<SyntaxTree, ImmutableArray<LoadDirective>> loadDirectiveMap,
+            ref ImmutableDictionary<SyntaxTree, Lazy<RootSingleNamespaceDeclaration>> declMap,
+            ref DeclarationTable declTable,
+            ref bool referenceDirectivesChanged)
+        {
+            // This HashSet will use default (case-sensitive) comparison.
+            // It is the responsibility of the SourceReferenceResolver to
+            // normalize the paths it resolves in a way that is appropriate
+            // for the platforms the host supports.
+            var seenFiles = PooledHashSet<string>.GetInstance();
+            AppendSyntaxTreesHelper(
+                treesBuilder,
+                tree,
+                tree.Options, // Use ParseOptions from external SyntaxTree for all sources it #loads.
+                this.Options,
+                seenFiles,
+                this.MessageProvider,
+                IsSubmission,
+                ordinalMapBuilder,
+                ref loadDirectiveMap,
+                ref declMap,
+                ref declTable,
+                ref referenceDirectivesChanged);
+            seenFiles.Free();
+        }
+
+        /// <summary>
+        /// Recursively adds trees #load'ed by <paramref name="tree"/> to the <paramref name="treesBuilder"/>.
+        /// </summary>
+        private static void AppendSyntaxTreesHelper(
+            ArrayBuilder<SyntaxTree> treesBuilder,
+            SyntaxTree tree,
+            ParseOptions parseOptions,
+            CSharpCompilationOptions options,
+            HashSet<string> seenFiles,
+            CommonMessageProvider messageProvider,
+            bool isSubmission,
+            IDictionary<SyntaxTree, int> ordinalMapBuilder,
+            ref ImmutableDictionary<SyntaxTree, ImmutableArray<LoadDirective>> loadDirectiveMap,
+            ref ImmutableDictionary<SyntaxTree, Lazy<RootSingleNamespaceDeclaration>> declMap,
+            ref DeclarationTable declTable,
+            ref bool referenceDirectivesChanged)
+        {
+            // It's okay if tree.FilePath is null or empty, because we only expect to see a single tree
+            // with no path information (the initial "code" passed to the CSharpScript constructor).
+            seenFiles.Add(tree.FilePath);
+
+            var resolver = options.SourceReferenceResolver;
+            ArrayBuilder<LoadDirective> loadDirectives = null;
+            foreach (var directive in tree.GetCompilationUnitRoot().GetLoadDirectives())
+            {
+                SyntaxTree loadedTree = null;
+                var diagnostics = DiagnosticBag.GetInstance();
+
+                var fileToken = directive.File;
+                var path = (string)fileToken.Value;
+                if (path == null)
+                {
+                    // If there is no path, the parser should have some Diagnostics to report.
+                    Debug.Assert(tree.GetDiagnostics().Any());
```

Comment:
Added check to assert (didn't find a "HasAnyErrors" that would work here, but passing a predicate that will check).